File: leetcode/20-valid-parentheses.py
# https://leetcode.com/problems/valid-parentheses/ (top-interview-questions)

class Solution:
	def isValid(self, s: str) -> bool:
		closing = {')': '(', ']': '[', '}': '{'}
		opening = list()
		for c in s:
			if c in closing:
				if len(opening) > 0 and closing[c] == opening[-1]:
					opening.pop()
				else: return False
			else: opening.append(c)

		return len(opening) == 0
	

	def is_valid(self, s: str) -> bool:
		opening = {'(': ')', '{': '}', '[': ']'}
		stack = list()
		for char in s:
			if char in opening:
				stack.append(char)
			elif not stack or opening[stack.pop()] != char:
				return False
		return len(stack) == 0

File: leetcode/20-valid-parentheses.js
// https://leetcode.com/problems/valid-parentheses/ (top-interview-questions)

/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function(s) {
  const closerParentheses = {')': '(', ']': '[', '}': '{'};
  const openedParentheses = []; // stack

  for (const char of s) {
    if (char in closerParentheses) {
      if (openedParentheses.length > 0 && closerParentheses[char] === openedParentheses[openedParentheses.length - 1]) {
        openedParentheses.pop();
      } else {
        return false;
      }
    } else {
      openedParentheses.push(char);
    }
  }

  return openedParentheses.length === 0;
};